Workflows CLI

MaxHerbs opened the pull request.

davehadley commented on workflows-cli/src/linter/helm.rs:
```diff
@@ -0,0 +1,120 @@
+    lint_from_manifest(&path_buf, true)
+}
+
+pub fn helm_template(target: &Path, template: Option<&str>) -> Result<Vec<String>, String> {
```
I think that all functions in this file except lint_from_helm can be private

davehadley commented on workflows-cli/src/linter/helm.rs:
```diff
@@ -0,0 +1,120 @@
+use crate::linter::base_linting::lint_from_manifest;
+
+pub fn lint_from_helm(target: &Path, all: bool) -> Result<Vec<LintResult>, String> {
+    let tmp_dir = Path::new("/tmp/argo-lint");
```
It might be easier and safer to use a "tempfile" / "tempdir" crate to handle these temp directory shenanigans for you.

MaxHerbs replied:
I'll have a look at some crates instead, that's a good idea

iamvigneshwars commented on workflows-cli/src/command_runner.rs:
```diff
@@ -0,0 +1,143 @@
+        let mappings: HashMap<String, Vec<MockEntry>> =
+            serde_yaml::from_str(&yaml_str).expect("Failed to parse YAML");
+
+        let target_mapping = env::var("RUST_TEST_ACTIVE_MAPPING").unwrap();
+        let active_mapping = mappings.get(&target_mapping).unwrap();
```
Is this safe? could we return `Self` as a `Result` ?

davehadley commented on workflows-cli/src/linter/mod.rs:
```diff
@@ -0,0 +1,212 @@
+    #[test]
+    #[serial]
+    fn lint_single_passing_manifest() {
+        unsafe {
```
Unsafe usage should usually be annotated with a safety comment.

iamvigneshwars commented on workflows-cli/src/main.rs:
```diff
@@ -0,0 +1,89 @@
+    }
+}
+
+use std::process::Command;
```
could we move this to the top?

iamvigneshwars commented on workflows-cli/tests/mock_commands.yaml:
EOF

iamvigneshwars commented on workflows-cli/src/command_runner.rs:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    /// Lookup command in the mapping table
+    fn get_response(&self) -> (String, i32) {
```
similarly could this be an `Option`?

iamvigneshwars commented on workflows-cli/src/linter/helm.rs:
```diff
@@ -0,0 +1,120 @@
+use crate::linter::base_linting::lint_from_manifest;
+
+pub fn lint_from_helm(target: &Path, all: bool) -> Result<Vec<LintResult>, String> {
+    let tmp_dir = Path::new("/tmp/argo-lint");
```
this could sometimes cause race conditions if multiple instances are running, we might have to use `tempfile` crate here

MaxHerbs replied:
Yes this is a good catch. I'll have a look into that

davehadley commented on workflows-cli/src/linter/mod.rs:
```diff
@@ -0,0 +1,212 @@
+    #[serial]
+    fn lint_single_passing_manifest() {
+        unsafe {
+            env::set_var("RUST_TEST_ACTIVE_MAPPING", "passing_manifest");
```
Why did you choose to modify the behavior of the MockCommand via environment variables? This seems to me that it might be potentially confusing to a future maintainer.

As a minimum, I would suggest adding comments to explain to a future maintainer what these environment variables are doing.

iamvigneshwars commented on workflows-cli/src/linter/helm.rs:
```diff
@@ -0,0 +1,120 @@
+
+    let select_template = match all {
+        true => None,
+        false => Some(target.file_name().unwrap().to_str().unwrap()),
```
is this safe?

davehadley commented on workflows-cli/src/command_runner.rs:
```diff
@@ -0,0 +1,143 @@
+        let mappings: HashMap<String, Vec<MockEntry>> =
+            serde_yaml::from_str(&yaml_str).expect("Failed to parse YAML");
+
+        let target_mapping = env::var("RUST_TEST_ACTIVE_MAPPING").unwrap();
```
It would be best to "namespace" any special environment variables that our application depends on.

eg WORKFLOWS_CLI_TEST_ACTIVE_MAPPING or similar

davehadley commented on workflows-cli/src/command_runner.rs:
```diff
@@ -0,0 +1,143 @@
+
+/// Get an instance of the CLI builder
+pub fn get_command_factory() -> Box<dyn CommandFactory> {
+    match env::var("MOCK_COMMAND") {
```
It would be best to "namespace" any special environment variables that our application depends on.

eg WORKFLOWS_CLI_TEST_ENABLE_MOCK_COMMAND or similar

davehadley replied:
Why did you choose to have the mocking live in the real application and not exist only in the test module?

Is the intention to have the ability to run the tests using the real "argo" and "helm" commands if they are installed on the system?

iamvigneshwars commented on workflows-cli/src/linter/config_linting.rs:
```diff
@@ -0,0 +1,61 @@
+}
+
+pub fn lint_from_config(config_path: &Path) {
+    let config = parse_config(config_path).unwrap_or_else(|e| {
+        eprintln!(
+            "Could not open config file {}: {}",
+            config_path.display(),
+            e
+        );
+        std::process::exit(1);
+    });
+
+    let mut lint_results = Vec::new();
+
+    if let Some(manifests) = config.manifests {
+        for manifest in &manifests {
+            match lint_from_manifest(manifest, true) {
+                Ok(mut results) => lint_results.append(&mut results),
+                Err(e) => {
+                    eprintln!("Error linting manifest '{}': {}", manifest.display(), e);
+                    std::process::exit(1);
+                }
+            }
+        }
+    }
+
+    if let Some(charts) = config.charts {
+        for chart in &charts {
+            match lint_from_helm(chart, true) {
+                Ok(mut results) => lint_results.append(&mut results),
+                Err(e) => {
+                    eprintln!("Error linting chart '{}': {}", chart.display(), e);
+                    std::process::exit(1);
+                }
+            }
+        }
+    }
+
+    print_and_exit(lint_results);
+}
```
could we decouple this logic from process termination? we could propogate the error back and process termination could happen at the top of the stack?

iamvigneshwars commented on workflows-cli/src/linter/mod.rs:
```diff
@@ -0,0 +1,212 @@
+    }
+}
+
+fn print_and_exit(results: Vec<LintResult>) {
+    results.iter().for_each(|result| {
+        print!("Template Name: {} - ", result.name);
+
+        if result.errors.is_empty() {
+            println!("No errors");
+        } else {
+            println!("Errors found");
+            result.errors.iter().for_each(|e| println!("  - {}", e));
+        }
+    });
+    match results.iter().any(|result| !result.errors.is_empty()) {
+        true => std::process::exit(1),
+        false => std::process::exit(0),
+    }
+}
```
could we decouple this logic from process termination? we could propogate the error back and process termination could happen at the top of the stack?

MaxHerbs commented on .github/workflows/_cli_code.yaml:
```diff
@@ -0,0 +1,125 @@
+
+      - name: Download tarpaulin
+        run: |
+          curl -L https://github.com/xd009642/tarpaulin/releases/latest/download/cargo-tarpaulin-x86_64-unknown-linux-gnu.tar.gz | tar xz
```
Installing with cargo builds the tarpaulin binary which can take ~2 minutes in CI. Tarpaulin then rebuilds everything so it can to do code coverage, which can take another 2/3 minutes. Downloading the binary instead is much quicker, and takes a second or two instead of minutes, which generally makes the processes faster.

MaxHerbs commented on workflows-cli/src/linter/helm.rs:
```diff
@@ -0,0 +1,120 @@
+
+    let target = match target.is_absolute() {
+        true => target,
+        false => &Path::new("./").join(target),
```
I've added the following comment:

```rust
    // When --all is provided, we traverse up the file path to find the directory with the chart.yaml
    // in it. If a user passes 'templates/workflow.yaml', this will check only the templates dir,
    // and doesnt check '.' as this is not explicitly in the path which will cause the lint to fail
    // if lint is called from the same directory as the Chart.yaml.
    // Appending './' adds '.' to the path traversal, which makes sure the working directory is also
    // checked for Chart.yaml
```

MaxHerbs commented on .github/workflows/_cli_code.yaml:
```diff
@@ -0,0 +1,125 @@
+        uses: actions/upload-artifact@v4
+        with:
+          name: coverage.html
+          path: workflows-cli/tarpaulin-report.html
```
It seems to work really badly with workspaces - I gave it a go, but it give one coverage for all of them, and includes the 80,000 lines of workflows structs that are created from the swagger schema so it comes out to like 1%. I only had a quick look so there might be config somewhere in it to handle that better

MaxHerbs commented on workflows-cli/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "workflows"
```
It would name the binary workflows-cli in that case - possibly doesnt matter? I've overwritten the package name in release please to workflows-cli for our clarity

iamvigneshwars replied:
Ahhh got it

MaxHerbs commented:
This should be ready for re-review. I've implimented most of the requested changes except;

I've left the tmpdir as is. Moving to a tempdir module was pretty easy, but this broke the tests as the path wasn't determinate anymore so I reverted that. I can spend a bit more time investigating that, but I dont see many cases where the race condition would surface as this will usually be run on personal workspaces, and the chances of a clash are pretty low.

More generally, I've added a more comments especially around the testing to clarify how that works for anyone working on this in future.

iamvigneshwars reviewed (approved):
LGTM
